[CUBRIDMAN-233] For UPDATE JOIN statements, aggregate functions or analytic functions cannot be used in the SET and WHERE clauses (#534)
CUBRID/cubrid-manual: en/sql/query/update.rst

UPDATE
******

You can update the column value of a record stored in the target table or view to a new one by using the **UPDATE** statement.

> Specify the name of the column to update and a new value in the **SET** clause, and specify the condition to be used to extract the record to be updated in the :ref:`where-clause`.

Reviewer: ```suggestion
Specify the name of the column to update and the new value in the **SET** clause, and specify the condition to extract the record to be updated in the :ref:`where-clause`.
```

Author: 감사합니다. 변경했습니다.

> You can one or more tables only with one **UPDATE** statement.

Author: 마침표를 기준으로 줄바꿈만 하였습니다. 웹페이지에서는 기존과 동일하게 한 줄로 표시됩니다.

> You can one or more tables only with one **UPDATE** statement.

Reviewer: ```suggestion
You can update one or more tables with a single **UPDATE** statement.
```

Author: 감사합니다. 변경했습니다.

*   *row_count*: Specifies the number of records to be updated after the :ref:`limit-clause`. It can be one of unsigned integer, a host variable or a simple expression.

> The following is allowed only when a single table is specified in <*table_specifications*>:

Reviewer: ```suggestion
The following are allowed only when a single table is specified in <*table_specifications*>:
```

Author: 감사합니다. 변경했습니다.

> * :ref:`order-by-clause` can be specified.

Reviewer: ```suggestion
* The :ref:`order-by-clause` can be specified.
```

Author: 감사합니다. 변경했습니다.

>   If :ref:`order-by-clause` is specified, records are updated in the order of the specified column.

Reviewer: ```suggestion
  If the :ref:`order-by-clause` is specified, records are updated in the order of the specified column.
```

Author: 감사합니다. 변경했습니다.

  This is useful for maintaining the order of trigger execution and the order of locking.

> * :ref:`limit-clause` can be specified.

Reviewer: ```suggestion
* The :ref:`limit-clause` can be specified.
```

Author: 감사합니다. 변경했습니다.

>   If :ref:`limit-clause` is specified, the number of records to be updated can be limited.

Reviewer: ```suggestion
  If the :ref:`limit-clause` is specified, the number of records to be updated can be limited.
```

Author: 감사합니다. 변경했습니다.

	           id  name                  phone
	=========================================================
	         NULL  NULL                  '000-0000'
	            4  NULL                  '000-0000'
	            5  NULL                  '000-0000'
	            7  NULL                  '777-7777'

    UPDATE t1 SET name='update', phone='999-9999' WHERE name IS NULL LIMIT 3;

    SELECT * FROM t1;

	           id  name                  phone
	=========================================================
	         NULL  'update'              '999-9999'
	            1  'aaa'                 '111-1111'
	            2  'bbb'                 '222-2222'
	            3  'ccc'                 '333-3333'
	            4  'update'              '999-9999'
	            5  'update'              '999-9999'
	            6  'ddd'                 '000-0000'
	            7  NULL                  '777-7777'

.. _example_update_with_joins:

.. rubric:: Example 3. Update Using Joins

When a record in table **A** is joined with multiple records in table **B** in an **UPDATE** statement,
the record in **A** is updated using only the value from the first matching record in **B**.

> In this example, a record in *t1* table with an *id* of 3 is joined with two records in *t2* table where *rate_id* is 3.

Reviewer: ```suggestion
In this example, a record in table *t1* with an *id* of 3 is joined with two records in table *t2* where *rate_id* is 3.
```

Author: 감사합니다. 변경했습니다.

    INSERT INTO t2 VALUES (2, 2);
    INSERT INTO t2 VALUES (3, 3);
    INSERT INTO t2 VALUES (4, 4);
    INSERT INTO t2 VALUES (6, 6);

    CREATE VIEW v1 AS
      SELECT b.id, b.val FROM t2 b LEFT JOIN t1 a ON b.id = a.id WHERE b.id <= 3;

    UPDATE v1 SET val = -1;

    SELECT * from v1;

	           id          val
	==========================
	            1           -1
	            2           -1
	            3           -1

    SELECT * from t2;

	           id          val
	==========================
	            1           -1
	            2           -1
	            3           -1
	            4            4
	            6            6

.. warning::
  
    Records in views that include tables with REUSE OID cannot be updated.

    For details, see :ref:`reuse-oid` and :ref:`dont-reuse-oid`.

.. _example_update_using_update_use_attribute_references:

.. rubric:: Example 5. Update Using the update_use_attribute_references Parameter

The result of this example depends on the value of the :ref:`update_use_attribute_references <update_use_attribute_references>` parameter.

*   If the value of this parameter is yes, *c2* is updated to 10, influenced by c1 = 10.

> *   If the value of this parameter is no, *c2* is not influenced by *c1 = 10* and is updated to 1 based on the value of c1 stored in the record.

Reviewer: ```suggestion
*   If the value of this parameter is no, *c2* is not influenced by *c1 = 10* and is updated to 1, based on the value of c1 stored in the record.
```

Author: 감사합니다. 변경했습니다.

.. code-block:: sql 

    DROP TABLE IF EXISTS t1;

    CREATE TABLE t1 (c1 INT, c2 INT); 
    INSERT INTO t1 values (1, NULL);

    SET SYSTEM PARAMETERS 'update_use_attribute_references=yes';

    UPDATE t1 SET c1 = 10, c2 = c1;

    SELECT * FROM t1;

	           c1           c2
	==========================
	           10           10

.. code-block:: sql

    TRUNCATE TABLE t1;
    INSERT INTO t1 values (1, NULL);

    SET SYSTEM PARAMETERS 'update_use_attribute_references=no';

    UPDATE t1 SET c1 = 10, c2 = c1;

    SELECT * FROM t1;

	           c1           c2
	==========================
	           10            1

.. _example_single_table_update_using_analytic_functions:

.. rubric:: Example 6. Single Table Update Using Analytical Functions

> When only one table is specified in the **UPDATE** statement, analytical functions can be used in the **SET** clause.

Reviewer: ```suggestion
When only one table is specified in the **UPDATE** statement, analytic functions can be used in the **SET** clause.
```

Author: 감사합니다. 변경했습니다.

.. code-block:: sql

    DROP TABLE IF EXISTS t1, t2;

    CREATE TABLE t1 (id INT);
    INSERT INTO t1 VALUES (1), (2), (3), (4), (5);

    CREATE TABLE t2 (id INT, val INT, update_val DOUBLE, join_update_val DOUBLE);
    INSERT INTO t2 (id, val) SELECT a.id, b.id FROM t1 a, t1 b WHERE b.id <= a.id;

    UPDATE t2 SET update_val = AVG (val) OVER (PARTITION BY id);

    SELECT DISTINCT id, TO_CHAR (update_val) AS update_val FROM t2;

	           id  update_val
	===================================
	            1  '1'
	            2  '1.5'
	            3  '2'
	            4  '2.5'
	            5  '3'

.. _example_multiple_tables_update_using_analytic_functions:

.. rubric:: Example 7. Multiple Tables Update Using Analytical Functions

> The example continues from :ref:`example_single_table_update_using_analytic_functions`.

Reviewer: ```suggestion
This example continues from :ref:`example_single_table_update_using_analytic_functions`.
```

Author: 감사합니다. 변경했습니다.

> when multiple tables are specified in the **UPDATE** statement, analytical functions cannot be used in the **SET** clause.

Reviewer: ```suggestion
When multiple tables are specified in the **UPDATE** statement, analytic functions cannot be used in the **SET** clause.
```

Author: 감사합니다. 변경했습니다.

> However, if a **SELECT** query is specified in the **SET** clause, analytical functions can be used within that **SELECT** query, regardless of the number of tables specified.

Reviewer: ```suggestion
However, if a **SELECT** query is specified in the **SET** clause, analytic functions can be used within that **SELECT** query, regardless of the number of tables specified.
```

Author: 감사합니다. 변경했습니다.

.. _example_local_table_update_using_joins_with_remote_tables:

.. rubric:: Example 9. Local Table Update Using Joins with Remote Tables

.. code-block:: sql

    /**
     * on the remote server (e.g., 192.168.6.21)
     */

    DROP TABLE IF EXISTS t2;

    CREATE TABLE t2 (rate_id INT, rate DOUBLE);

.. code-block:: sql

    /**
     * on the local server (e.g., 192.168.6.22)
     */

    -- DROP SERVER remote;
    CREATE SERVER remote (HOST='192.168.6.21', PORT=33000, DBNAME=demodb, USER=public);

    DROP TABLE IF EXISTS t1;

    CREATE TABLE t1 (id INT PRIMARY KEY, charge DOUBLE);
    INSERT INTO t1 VALUES (1, 100.0);
    INSERT INTO t1 VALUES (2, 100.0);
    INSERT INTO t1 VALUES (3, 100.0);

    INSERT INTO t2@remote VALUES (1, 0.1);
    INSERT INTO t2@remote VALUES (2, 0.2);
    INSERT INTO t2@remote VALUES (3, 0.3);
    INSERT INTO t2@remote VALUES (3, 0.5);

    UPDATE t1 a INNER JOIN t2@remote b ON a.id = b.rate_id
    SET a.charge = a.charge * (1 + b.rate);

    SELECT id, TO_CHAR (charge) AS charge FROM t1;

	           id  charge
	===================================
	            1  '110'
	            2  '120'
	            3  '150'

.. warning::

>     In the **UPDATE** statements that join local and remote tables, records in the remote table cannot be updated.

Reviewer: ```suggestion
    In **UPDATE** statements that join local and remote tables, records in the remote table cannot be updated.
```

Author: 감사합니다. 변경했습니다.

비슷한 변경이 필요한 것들도 함께 변경하고 싶었지만 대상을 찾기가 어려웠습니다.
전체적으로 한 번 더 리뷰해주시면 감사하겠습니다.
